Infer node types by following definitions

richard0000 opened the pull request:
# What was done?

## Infer node types by following definitions

### Improve the user experience on uploading specs with failures

- Warn the user to select a domain if not selected
- Fail with a toast error message if not classes found.
- Include JSON Schema converter to JSON-LD

### Build JsonLd parser to infer node types

In Json-ld the type of a node may be expressed in different ways. It could
be a full HTTP URI, a prefixed URI, or it may not be typed. In order to
recognize classes and properties, it may be necessary to fetch resources on
the internet.

Now we can do that for each node. It implies a performance impact, but it's
planned to incorporate caching before it's merged because many nodes of the
now the same type are being analyzed repeatedly in separate requests, because
we don't have chacing at the moment.

### Refactor specifications processor

To be consistent with the new JsonLd parser, make the processor
not static and place methods in the proper classes.

### Implement cache using Redis to save node types

Many node types might be repeated, so we cache them in order to minimize
the HTTP requests.

excelsior commented on app/services/converters/json_schema.rb:
```diff
@@ -0,0 +1,21 @@
+    # @param file [#path]
+    # @return [Hash]
+    def self.convert(file)
+      reader = RDF::JSON::Reader.open(file.path)
```
This won't work, because `RDF::JSON::Reader` parses RDF/JSON data, not JSON Schema. JSON Schema is covered by the `JsonSchemaZip` converter. If individual JSON Schema uploads are needed, we better extract that functionality from `JsonSchemaZip`.

Also, this particular thing seems to be out of the scope of the node type inference feature.

richard0000 replied:
@excelsior Ok for the first part.

For the second and last comment, I would like to keep it here. I observed that the JSON file used in Issue #79 is not in the format the application needs. But after this processing, it changes to be the exact expected format.

excelsior replied:
Ah, that file is in the RDF/JSON format indeed. Just rename the converter to `RdfJson` to avoid confusion.

excelsior reviewed (changes requested):
I tested the feature with the ASN files provided by Stuart and couldn't quite make it work.

AFAIK that spec contains only two types: `asn:StandardDocument` and `asn:Statement` (both of which are RDF classes according to the context). So I expected those two being listed in the multiple domains modal. However, more than 200 domains with repeated names were displayed (see screenshot), whose URIs indicated actual resources and not classes. I also tried uploading the Darwin Core spec (which contains actual classes), and it worked as supposed.

<img width="1440" alt="Screen Shot 2021-01-13 at 13 35 03" src="https://user-images.githubusercontent.com/406292/104429346-5510dd00-55a7-11eb-83a4-e7fdd3153d19.png">

richard0000 commented on app/javascript/components/mapping/FileInfo.jsx:
```diff
@@ -23,7 +23,7 @@ const FileInfo = (props) => {
         </p>
         <p>
-          <strong>File Type:</strong> {mime.getType(selectedFile.name)}
+          <strong>File Type:</strong> {selectedFile.type}
```
@excelsior `mime.getType` fails with error:
`Uncaught TypeError: mime__WEBPACK_IMPORTED_MODULE_0___default.a.getType is not a function`

![image](https://user-images.githubusercontent.com/6996951/104454052-0ef65200-5584-11eb-8bc6-dda272308a84.png)

excelsior replied:
It works on staging, so something might be wrong with your local setup.

excelsior commented:
One more suggestion: let's keep the type information inferred from the contexts in the DB instead of Redis.

richard0000 commented:
> One more suggestion: let's keep the type of information inferred from the contexts in the DB instead of Redis.

@stuartasutton I've added a fix to this issue, and now is under review.

BTW, the last provided file (ASN Standard) had an issue with the `dcterms` prefix. I removed lines 24 and 25 and now we can consume the file. See screenshot below:

![rdf-processing](https://user-images.githubusercontent.com/6996951/104639857-a6dc6480-5686-11eb-9f1e-3c0a2cd7a2ca.gif)

excelsior commented on db/migrate/20210113213459_create_rdfs_class_nodes.rb:
```diff
@@ -0,0 +1,12 @@
+class CreateRdfsClassNodes < ActiveRecord::Migration[6.0]
+  def change
+    create_table :rdfs_class_nodes do |t|
+      t.string :uri
```
This column needs a unique index.

richard0000 replied:
Reviewing this and previous comments @excelsior

excelsior reviewed (changes requested):
Thanks Ricardo,

Type inference works as expected, congrats. However, filtration doesn't seem to function as expected: when I select a type, the preview displays an empty graph (the Darwin Core file still works, tho). Moreover, non-RDF uploads started failing during the class retrieval phase. We need to ensure compatibility of those formats with the new flow.
